spec.md

```diff
@@ -83,28 +113,55 @@ $\operatorname{SPLIT}_C \in V_8 \rightarrow V_v$
 
 The configuration must satisfy $S_{\text{max}} \ge S_{\text{min}} > 0$.
 
+<!---
+NOTE
+
+It might help the clarity of what follows
+if we make parameterization by C implicit
+rather than repeating C in subscripts everywhere.
+-->
+
 ## Definitions
 
 We define the constant $W$, which we call the "window size," to be 64.
 
-The "split index" $I(X)$ of a sequence $X$ is either the smallest
-non-negative integer $i$ satisfying:
+<!---
+NOTE
+
+Fixing W at 64
+arbitrarily rules out using this section
+to describe and reason about hashsplit algorithms
+that it otherwise could.
 
-- $i \le |X|$ and
-- $S_{\text{max}} \ge i \ge S_{\text{min}}$ and
-- $H(\langle X_{i-W}, \dots, X_{i-1} \rangle) \mod 2^T = 0$
+I think fixing it at 64 is more properly a part of the recommendation section.
```

Comment:
I really do want to be prescriptive about this, because fixing it as a constant allows implementations that aren't possible if this is configurable. In particular, this allows using a fixed size array for the ring buffer rather than a dynamically allocated slice (or your language's equivalent). For e.g. @cole-miller's Rust implementation this is a big deal, because it means rolling hashes can be computed in an environment that doesn't supply a heap allocator, and generally it will be more efficient.

I guess in my mind if something is "recommended" then general purpose libraries should still allow a user to configure it, as they may need to to interoperate with other systems.

Reply:
Just to clarify -- I'm already using Rust's [nascent support](https://github.com/rust-lang/rust/issues/74878) for const generics in my implementation, and it would be pretty simple to make `WINDOW_SIZE` a generic parameter in the appropriate place. That would allow users to choose a custom window size, as long as it's specified at compile time (which is a pretty mild restriction). The general point about a fixed window size making things easier for implementors seems reasonable, though.

A compromise that seems reasonable to me is to explicitly acknowledge the possibility of using other window sizes, while making clear that this is a "bonus feature" that spec-compliant implementations are not required to support.

Reply:
Hi! Thanks for the comments and sorry for the delay.

In the end I will defer to your preference on this. But before I do I want to push back a bit more. What is the purpose of the formalism in this part of the document? I thought the whole point was to describe essentially the whole space of possible hashsplit algorithms (or at least, all those belonging to a broad family) simply by the selection of appropriate configuration parameters.

This allows formal reasoning about a wide assortment of different hashsplit algorithms, and to that end I see no use in artificially limiting the size of that assortment by removing a parameter. Keeping `W` as a configurable value neither complicates the description nor limits our ability to prescribe setting it to 64.

On the other hand, if the only purpose of this document is to prescribe a specific hashsplit algorithm (and perhaps to describe one or two others, like rrs), why bother with the formalism at all? Why not describe exactly and only those algorithms and no others?

(These same argument applies to the input-shorter-than-W business below, unless we're re-imposing the S_min>=W requirement, in which case it's moot.)

Reply:
I don't have a problem with making W an opaque parameter from the perspective of the formal description. I suppose part of the trouble is we don't really have a section on recommendations for implementors, and so I'd been thinking of the configuration as much as a set of parameters that would be exposed by a library as part of the formalism. I think it makes sense to expose S_min, S_max and T to library users, I'm waffling on H, and I think W does not, so maybe the right answer is to actually separate these two ideas, and add a separate section that is prescriptive about what conforming library implementations should or should not make configurable. I'll open a new issue about this.

Reply:
> maybe the right answer is to actually separate these two ideas, and add a separate section that is prescriptive about what conforming library implementations should or should not make configurable

I've been assuming all along that that was the plan. Sorry for the miscommunication, but I do think that's the right way to go.